Add `fde-custom` feature

nspin opened the pull request:
This PR proposes a feature called `fde-deferred` which, when selected, enables a new FDE finding strategy. This strategy depends on the following symbols, which may be defined by dependent crates:

```rust
extern "Rust" {
    // Required
    fn __unwinding_get_text_base(pc: usize) -> Option<usize>;

    // At least one of these must be defined
    fn __unwinding_get_eh_frame_hdr() -> Option<usize>;
    fn __unwinding_get_eh_frame() -> Option<usize>;
}
```

This strategy provides maximal flexibility, at the expense of requiring support from dependent crates. That is, the discovery of these base addresses is deferred to dependent crates.

As a concrete example of the usefulness of a feature like this, I'm using this crate in a bare microkernel userspace environment. In this instance, a program's PHDRs are made available to it in a nonstandard way. This `fde-deferred` feature would allow dependent runtime crates to pass the relevant base address information back to `unwinding`. In this case, for example, `fde-deferred` has the advantage of being linker-independent (unlike `fde-gnu-eh-frame-hdr`) and does not require a custom linker script (unlike `fde-static`).

If you are considering accepting this PR, then I will, of course, add documentation.

nspin committed:
Add fde-custom feature

nbdd0121 commented:
I wonder if I should extract things to an extra trait and then use a mechanism adds API to register FDE finders..

nspin commented:
That would probably be a cleaner solution.

I would propose the following constraint on such a solution, however, in order to maximize its applicability: that it not depend on `alloc`.

nbdd0121 commented:
Yeah, that's always the goal for this project. I spent a lot of effort to make gimli usable without liballoc.

nbdd0121 commented:
I think one approach is to have a trait and then add a macro, given a static instance of the trait, convert it to `extern "Rust"` functions. E.g.
```rust
struct CustomEhFrameFinder;

impl BikesheddingTraitName for CustomEhFrameFinder { ... }

unwinding::custom_eh_frame_finder! {
    static CUSTOM: CustomEhFrameFinder = CustomEhFrameFinder;
}
// this expands to
#[no_mangle]
fn __unwinding_get_eh_frame_hdr() -> Option<usize> {
    CUSTOM....
}
```
this way we can keep the symbol name internal to the crate (so not part of the API) and make it easier to use and define (and we can avoid adding linkage feature).

nspin commented:
> Yeah, that's always the goal for this project. I spent a lot of effort to make gimli usable without liballoc.

And I appreciate that effort very much!

> I think one approach is to have a trait and then add a macro, given a static instance of the trait, convert it to `extern "Rust"` functions. E.g...

Another approach is one similar to that found in the [`log`](https://docs.rs/log/latest/log/) crate. Inside the crate, there is:

```rust
static mut LOGGER: &dyn Log = ...;
```

On platforms with atomics, it is protected with some simple atomic-based synchronization.

The API provides different ways of setting the global logger depending on what is available:
- `pub fn set_boxed_logger(logger: Box<dyn Log>)` for when atomics and `alloc` are available (implemented with `Box::leak`)
- `pub fn set_logger(logger: &'static dyn Log)` for when atomics are available
- `pub unsafe fn set_logger_racy(logger: &'static dyn Log)` for when atomics are not available

nspin commented:
I've updated this PR to follow an approach similar to that of the `log` crate. I've also changed the name of the proposed feature to `fde-custom`.

If you like this approach, I'll go ahead and add documentation (rustdoc for new public items + an addition to the FDE section of the README).

nspin commented on src/lib.rs:
```diff
@@ -20,6 +20,14 @@ extern crate alloc;
 #[cfg(feature = "unwinder")]
 mod unwinder;
 
+#[cfg(all(feature = "unwinder", feature = "fde-custom"))]
+pub mod custom_fde_finder {
+    pub use super::unwinder::{
+        set_custom_fde_finder, BaseAddressFinder, FDEFinder, FDESearchResult,
+        SetCustomFDEFinderError,
+    };
+}
```
How would you like these items, now necessarily public, to be exposed?

nbdd0121 replied:
I don't really want to expose the internals. Maybe instead of custom "FDEFinder", it could be a custom "EhFrameFinder" (which you currently call `BaseAddressFinder`) that is registered? The whole FDE finding mechanism can still be internal.

nspin replied:
That makes sense. My original idea was to allow for maximal flexibility (that is, to allow dependencies to set global finders which are not based on `BaseAddressFinder`, but I don't think this degree of flexibility would ever be necessary.

Done.

nbdd0121 commented on src/unwinder/find_fde/custom.rs:
```diff
@@ -0,0 +1,135 @@
+    fde_finder: &'static (dyn FDEFinder + Sync),
+) -> Result<(), SetCustomFDEFinderError> {
+    if CUSTOM_FDE_FINDER_INIT_STARTED.swap(true, Ordering::SeqCst) {
+        Err(SetCustomFDEFinderError(()))
```
This is problematic because it can cause a case where custom FDE finder is not yet set but `Err` is returned. It should spin until the custom fde finder is set, and then return `Err`.

nspin replied:
Fixed.

nbdd0121 commented on src/unwinder/find_fde/custom.rs:
```diff
@@ -0,0 +1,135 @@
+}
+
+pub trait BaseAddressFinder {
+    fn find_text_base(&self, _pc: usize) -> Option<usize>;
```
How about
```rust
enum FrameInfoKind {
    EhFrameHdr(usize),
    EhFrame(usize),
}

struct FrameInfo {
    pub text_base: usize,
    pub kind: FrameInfoKind,
}
```

and a single method finding `FrameInfo` from PC? BTW this trait should be unsafe to implement.

nspin replied:
Good idea.

Done.

nbdd0121 reviewed (commented):
Thanks, this version looks much cleaner. Would you mind change the two `AtomicBool` initialization to a single `AtomicU32` instead? You can follow the approach used in `log` crate. Thanks!
